Use flyway for migrations

egh opened the pull request:
https://jira.plos.org/jira/browse/AMBR-102

I tried out flyway, and it proved pretty easy to use.

I used a sql schema dump from production with the `versions` table (used by the old db migrations) removed and the auto increments not set.

Start from zero tables:
```
λ echo "show tables;" |mysql -u root -p ambra 
Enter password: 
[nothing printed]
```
Start tomcat:

```λ mvn tomcat7:run -Drhino.configDir=/home/egh/c/ambra/config/
[INFO] Scanning for projects...
[...]
Sep 27, 2018 1:56:11 PM org.apache.coyote.AbstractProtocol start
INFO: Starting ProtocolHandler ["http-bio-8080"]
```

Check tables:
```
λ echo "show tables;" |mysql -u root -p ambra 
Enter password: 
Tables_in_ambra
annotation
annotationCitation
annotationCitationAuthor
annotationCitationCollabAuthor
annotationFlag
article
articleAsset
articleCategoryAssignment
articleCategoryAssignmentFlag
articleCategoryFlagged
articleCategoryJoinTable
articleCollaborativeAuthors
articleFile
articleIngestion
articleItem
articleList
articleListJoinTable
articlePerson
articlePublishedJournals
articleRelationship
articleRevision
articleType
category
categoryFeaturedArticle
citedArticle
citedArticleCollaborativeAuthors
citedPerson
comment
commentFlag
flyway_schema_history
issue
issueArticleList
jira_ambr_291
journal
oldArticle
oldArticleList
oldArticleListJoinTable
oldArticleRelationship
oldIssue
oldIssueArticleList
oldJournal
oldSyndication
oldVolume
pingback
savedSearch
savedSearchQuery
syndication
trackback
userArticleView
userLogin
userOrcid
userProfile
userProfileMetaData
userProfileRoleJoinTable
userRole
userRolePermission
userSearch
volume
```

slimeate commented:
This is awesome and I feel like a dummy for not being able to figure it out with plosmatch. 

What's going to happen when this deploys to a server that already has the schema set up (like prod)? I'd be afraid that flyway would try and fail to run the v1 migration because there's no preexisting flyway_schema_history table and that would prevent tomcat from starting.

egh commented:
@slimeate Flyway has a `baseline` command that you run in exactly that case.

slimeate commented:
> @slimeate Flyway has a `baseline` command that you run in exactly that case.

Do we have to manually run it or is that automatically run by https://github.com/PLOS/rhino/pull/458/files#diff-f29c9a6a6cafd81ed4384f85b90aa305R94 ?

egh commented:
@slimeate We'll need to run it manually, I believe.

openAccess commented:
Where did that db schema come from? I hope that is not in production.  I am fine with flyaway - I think we have discussed it a couple of time before.

egh commented:
@openAccess That is the schema dump of the production db, minus the `versions` table. Not sure what you mean by "I hope that is not in production"

openAccess commented:
Because if you look hard enough you will find that are several tables that are no longer used. People swore to me they had been deleted etc.

________________________________
From: Erik Hetzner <notifications@github.com>
Sent: Wednesday, October 3, 2018 10:33:57 AM
To: PLOS/rhino
Cc: Bill Oconnor; Mention
Subject: Re: [PLOS/rhino] Use flyway for migrations (#458)


@openAccess<https://github.com/openAccess> That is the schema dump of the production db, minus the versions table. Not sure what you mean by "I hope that is not in production"

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/PLOS/rhino/pull/458#issuecomment-426727144>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABOTdZG4eYnls0CLEpsJCbvlHRxjeRH8ks5uhPUEgaJpZM4W9cGm>.

egh commented:
@openAccess They are still in prod.

slimeate commented on src/main/webapp/WEB-INF/spring/appServlet/servlet-context.xml:
```diff
@@ -89,7 +89,12 @@
   </mvc:interceptors>
 
+  <beans:bean id="flyway" class="org.flywaydb.core.Flyway" init-method="migrate">
+    <beans:property name="locations" value="classpath:migrations/" />
```
This value defaults to  "classpath:db/migration". Since you're moving the migration files anyways, you might as well just move them to the conventional location.

slimeate commented:
> Satisfy Jack and then :shipit:

We talked about it yesterday in standup, @egh, but so it's on the record, I think we need to explicitly list out whatever manual steps are involved in deploying this to an existing database (ie prod), and optimally have that procedure tested in vagrant. If we merge this as-is and forget there's the manual baseline step involved, then the next release will fail.

On the topic of the unused prod tables, I agree with @sgatchell, that sounds like a separate ticket. For now, I'm of the opinion that whatever schema we have in prod should be our source of truth, which is how you currently have this. If it turns out there are indeed extra tables in that schema, then this fancy migration system you're adding should make removing them a breeze.

However, I recall you mentioning that the prod schema is not the same as the other stages. How do you think we should handle that? If everything is starting off with at different schema baseline, then the migrations probably wont run predictably and we'll lose our ability to simulate migrations of prod on lower-tier environments. We might be able to just wipe those dbs and start from scratch if we aren't attached to the data on them.

egh commented:
@slimeate I amended the existing commit and rebased on top of develop with the following changes:
- Used the default db migration location (thanks, I didn't realize I was overriding a default)
- Removed the possibly dangerous `baselineOnMigrate` options (https://flywaydb.org/documentation/commandline/migrate#baselineOnMigrate)
- Restored the old `version` table in the baseline sql migration. This was used by the old migration system, and I initially deleted it from the baseline, but I realize now that we need to ensure that the baseline is exactly the same as the existing prod db schema. We can remove this table later, properly, in a migration.

slimeate commented:
@egh, that all sounds good. I'm still curious though how we intend on manually baselining existing schemas when we deploy.

I'm not sure if you saw, but I duplicated your approach here in the aematch deb-packaging PR (so I could get rid of that scary schema load in the plosmatch-formulas PR). https://github.com/PLOS/aematching/pull/418/commits/8f083c757fed8e434c2e9d7e94692ddd7a4e0c23 I pretty much copied over what you did here. But I also have the same issue I'll need to deal with, of initializing this system on our prod db, so I have a very personal stake in figuring out how to do that.

slimeate commented:
. . . it appears to me that unless we use that `baselineOnMigrate` option, we'll need to make the flyway command-line tool available, which adds an additional layer of work here. It might just be easiest to use `baselineOnMigrate` when we first roll this out, then to set it to false before the subsequent release, after we know the legacy DBs have been migrated.

egh commented:
@slimeate I re-enabled `baselineOnMigrate`.

egh commented:
https://jira.plos.org/jira/browse/AMBR-700

slimeate commented:
:shipit:

egh commented:
Thank you @slimeate and @sgatchell !!
